plugins/base/src/main/kotlin/resolvers/local/DefaultLocationProvider.kt

```diff
@@ -0,0 +1,219 @@
+
+    override fun ancestors(node: PageNode): List<PageNode> =
+        generateSequence(node) { it.parent() }.toList()
+
+    protected open fun pathTo(node: PageNode, context: PageNode?): String {
+        fun pathFor(page: PageNode) = pathsIndex[page] ?: throw AssertionError(
+            "${page::class.simpleName}(${page.name}) does not belong to current page graph so it is impossible to compute its path"
+        )
+
+        val contextNode =
+            if (context?.children?.isEmpty() == true && context.parent() != null) context.parent() else context
+        val nodePath = pathFor(node)
+        val contextPath = contextNode?.let { pathFor(it) }.orEmpty()
+
+        val commonPathElements = nodePath.asSequence().zip(contextPath.asSequence())
+            .takeWhile { (a, b) -> a == b }.count()
+
+        return (List(contextPath.size - commonPathElements) { ".." } + nodePath.drop(commonPathElements) +
+                if (node.children.isNotEmpty()) listOf(PAGE_WITH_CHILDREN_SUFFIX) else emptyList()).joinToString("/")
+    }
+
+    private fun PageNode.parent() = pageGraphRoot.parentMap[this]
+
+
+
+
+    private val cache: MutableMap<URL, LocationInfo> = mutableMapOf()
+
+    private fun getLocation(dri: DRI, jdkToExternalDocumentationLinks: Map<Int, List<DokkaConfiguration.ExternalDocumentationLink>>): String {
+        val toResolve: MutableMap<Int, MutableList<DokkaConfiguration.ExternalDocumentationLink>> = mutableMapOf()
+        for((jdk, links) in jdkToExternalDocumentationLinks) {
+            for(link in links) {
+                val info = cache[link.packageListUrl]
+                if(info == null) {
+                    toResolve.getOrPut(jdk) { mutableListOf(link) }.add(link)
+                } else if(info.packages.contains(dri.packageName)) {
+                    return link.url.toExternalForm() + getLink(
+                        dri,
+                        info
+                    )
+                }
+            }
+        }
+        // Not in cache, resolve packageLists
+        for ((jdk, links) in toResolve.entries) {
+            for(link in links) {
+                val locationInfo =
+                    loadPackageList(
+                        jdk,
+                        link.packageListUrl
+                    )
+                if(locationInfo.packages.contains(dri.packageName)) {
+                    return link.url.toExternalForm() + getLink(
+                        dri,
+                        locationInfo
+                    )
+                }
+            }
+            toResolve.remove(jdk)
+        }
+        return ""
+    }
+
+    private fun getLink(dri: DRI, locationInfo: LocationInfo): String  =
+        locationInfo.locations[dri.packageName + "." + dri.classNames] ?: // Not sure if it can be here, previously it shadowed only kotlin/dokka related sources, here it shadows both dokka/javadoc, cause I cannot distinguish what LocationProvider has been hypothetically chosen
```

Comment:
Before they were distinguished this way in org.jetbrains.dokka.base.resolvers.ExternalLocationProvider
```
when(locationInfo.format) {
        "javadoc" ->  dri.toJavadocLocation(8)
        "kotlin-website-html", "html" -> locationInfo.locations[dri.packageName + "." + dri.classNames] ?: dri.toDokkaLocation(".html")
        "markdown" -> locationInfo.locations[dri.packageName + "." + dri.classNames] ?: dri.toDokkaLocation(".md")
        // TODO: rework this
        else -> throw RuntimeException("Unrecognized format")
    }
```
Right now I delegated it to dedicated registered `Extension Points` with `ExternalLocationProvider` factories. Factories have API, which does not include `LocationInfo` parameter particular exception for markdown. However `locations` property should be empty for other formats. Do you know any precautions when `locations` property can be not null and shadow resolving link with resolver?